### ISSUE TITLE
context.WithFactory(lambdaFunc) fails when using FastExpressionCompiler

### ISSUE THREAD

**Issue by schuettecarsten** [z4kn4fein/stashbox#71]
There is a NuGet library called "FastExpressionCompiler" to compile expression trees into faster Funcs. When passing such Funcs into Stashbox RegistrationConfigurator.WithFactory method, an exception is thrown when the context initializes. This exception occurs regardless if WithMicrosoftExpressionCompiler is used or not.

```
MethodInfo muss ein MethodInfo-Laufzeitobjekt sein.
Parametername: meth
   at System.Reflection.Emit.DynamicILGenerator.Emit(OpCode opcode, MethodInfo meth)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitMethodCall(MethodInfo mi, IArgumentProvider args, Type objectType, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitMethodCall(Expression obj, MethodInfo method, IArgumentProvider methodCallExpr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitMethodCallExpression(Expression expr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitConvert(UnaryExpression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitVariableAssignment(BinaryExpression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitAssign(BinaryExpression node, CompilationFlags emitAs)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpressionAsVoid(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.Emit(BlockExpression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitBlockExpression(Expression expr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitVariableAssignment(BinaryExpression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitAssign(BinaryExpression node, CompilationFlags emitAs)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpressionAsVoid(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.Emit(BlockExpression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitBlockExpression(Expression expr, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitExpression(Expression node, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitLambdaBody(CompilerScope parent, Boolean inlined, CompilationFlags flags)
   at System.Linq.Expressions.Compiler.LambdaCompiler.EmitLambdaBody()
   at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda, DebugInfoGenerator debugInfoGenerator)
   at System.Linq.Expressions.ExpressionExtensions.CompileDelegate(Expression expression, ResolutionContext resolutionContext, ContainerConfiguration containerConfiguration)
   at Stashbox.Lifetime.SingletonLifetime.GetExpression(IContainerContext containerContext, IServiceRegistration serviceRegistration, IObjectBuilder objectBuilder, ResolutionContext resolutionContext, Type resolveType)
   at Stashbox.Registration.ServiceRegistration.ConstructExpression(IContainerContext containerContext, ResolutionContext resolutionContext, Type resolveType)
   at Stashbox.Registration.ServiceRegistration.GetExpression(IContainerContext containerContext, ResolutionContext resolutionContext, Type resolveType)
   at Stashbox.ResolutionScope.Activate(ResolutionContext resolutionContext, Type type, Object name)
   at Stashbox.ResolutionScope.Resolve(Type typeFrom, Boolean nullResultAllowed, Object[] dependencyOverrides)
   at Stashbox.StashboxContainer.Resolve(Type typeFrom, Boolean nullResultAllowed, Object[] dependencyOverrides)
   at Stashbox.DependencyResolverExtensions.Resolve[TKey](IDependencyResolver resolver, Boolean nullResultAllowed, Object[] dependencyOverrides)
   ...
```

```
MethodInfo muss ein MethodInfo-Laufzeitobjekt sein.
Parametername: meth
   at System.Reflection.Emit.DynamicILGenerator.Emit(OpCode opcode, MethodInfo meth)
   at System.Reflection.Emit.ILGeneratorExtensions.EmitMethod(ILGenerator generator, MethodInfo info)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(MethodCallExpression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(Expression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(UnaryExpression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(Expression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(BinaryExpression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(Expression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(IList`1 expressions, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(Expression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(BinaryExpression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(Expression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(IList`1 expressions, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.Emitters.Emitter.TryEmit(Expression expression, ILGenerator generator, CompilerContext context, ParameterExpression[] parameters)
   at Stashbox.BuildUp.Expressions.Compile.ExpressionEmitter.TryEmit(Expression expression, Delegate& resultDelegate, Type delegateType, Type returnType, ParameterExpression[] parameters)
   at System.Linq.Expressions.ExpressionExtensions.CompileDelegate(Expression expression, ResolutionContext resolutionContext, ContainerConfiguration containerConfiguration)
   at Stashbox.Lifetime.SingletonLifetime.GetExpression(IContainerContext containerContext, IServiceRegistration serviceRegistration, IObjectBuilder objectBuilder, ResolutionContext resolutionContext, Type resolveType)
   at Stashbox.Registration.ServiceRegistration.ConstructExpression(IContainerContext containerContext, ResolutionContext resolutionContext, Type resolveType)
   at Stashbox.Registration.ServiceRegistration.GetExpression(IContainerContext containerContext, ResolutionContext resolutionContext, Type resolveType)
   at Stashbox.ResolutionScope.Activate(ResolutionContext resolutionContext, Type type, Object name)
   at Stashbox.ResolutionScope.Resolve(Type typeFrom, Boolean nullResultAllowed, Object[] dependencyOverrides)
   at Stashbox.StashboxContainer.Resolve(Type typeFrom, Boolean nullResultAllowed, Object[] dependencyOverrides)
   at Stashbox.DependencyResolverExtensions.Resolve[TKey](IDependencyResolver resolver, Boolean nullResultAllowed, Object[] dependencyOverrides)
   ...
```

I am not sure if this is a Stashbox or a FastExpressionCompiler issue, but it would be great if Stashbox would work with compiled expressions generated by FastExpressionCompiler. I use the FastExpressionCompiler for several other things without any issues.

Sample code:

```
internal static void ConfigureContextForSingleton(RegistrationConfigurator context, PropertyInfo propertyInfo)
{
	if (propertyInfo != null)
	{
		var memberExpr = Expression.MakeMemberAccess(null, propertyInfo);
		var lambdaExpr = Expression.Lambda<Func<object>>(memberExpr);
		var lambdaFunc = lambdaExpr.CompileFast();

		context
			.WithFactory(lambdaFunc)
			.WithSingletonLifetime()
			.WithoutDisposalTracking();
	}
	else
	{
		context.AsConstant(null);
	}
}
```

It works when using the standard "lambdaExpr.Compile()" method, but fails with CompileFast().

**Comment by z4kn4fein** [comment]
Thanks for reporting! I'll check this out!

**Comment by z4kn4fein** [comment]
I didn't find a good way to determine that the passed delegate was compiled with a custom Expression compiler. It's possible to tell that it was compiled with Microsoft's default expression compiler, but that's all. So I decided to let the user tell that it's a compiled delegate and the container will handle it.

So when you configure your registration you can pass a `true` as the second parameter, which indicates that the passed delegate is a compiled one:
`c.WithFactory(memberLambda, true)`

You can use this feature by using the latest [pre-release package](https://www.myget.org/feed/pcsajtai/package/nuget/Stashbox). Please let me know if you have further questions or suggestions!

**Comment by schuettecarsten** [comment]
Thank you very much. This works as expected.

**Comment by z4kn4fein** [comment]
released in [v2.8.6](https://github.com/z4kn4fein/stashbox/releases/tag/2.8.6)
